src/ngraph/op/util/attr_types.hpp

```diff
@@ -166,7 +166,8 @@ namespace ngraph
             NONE = 0,
             EXPLICIT = NONE,
             NUMPY,
-            PDPD
+            PDPD,
+            BIDIRECTIONAL
```

Comment:
What will happen if someone will use BIDIRECTIONAL mode for v1::Broadcast?
We should probably detect this situation to avoid undocumented behavior.

Reply:
done, check added

Reply:
Our original specification for Broadcast-1 was misleading:

 "If `mode` is set to `numpy`, it     means that the standard one-directional numpy broadcasting rules are applied. They are similar to rules that applied in all binary element-wise operations in case when `auto_broadcasting` attribute is set to `numpy`, and are similar to rules described at [here](https:/    /docs.scipy.org/doc/numpy/user/basics.broadcasting.html#general-broadcasting-rules), when only one-directional broadcasting is applied: input tensor `data` is broadcasted to `target_shape` but not vice-versa."

I think that both UNIDIRECTIONAL and BIDIRECTIONAL will have the same level of confusion for people who are familiar with broadcasting rules.

What I currently see in the nGraph op code is that the broadcasting with NUMPY mode is a kind of mix of uni- and bi-directional broadcasting. If the length of the target_shape is less than the length of then input shape we raise an exception. At the same time when we compare respective dimensions of the input tensor and the target shape we check that they are equal or one of them is equal to 1. This is the behaviour of bi-directional broadcasting. This means that

Reply:
Oh, I now see the awkwardness of my comment: we cannot broadcast the tensor to smaller shape.

Reply:
Now I see that the shape infer function of Broadcast with NUMPY mode works not a described in the documentation. According to the specification if the input tensor shape is (10,1) and target shape is (1,20) then error should be produced, but the implementation will produce output shape (10,20). But the output shape (10,20) should be produced only for BIDIRECTIONAL model.

@arogowie-intel, do you agree?

Reply:
@arogowie-intel To be clear, as I understand numpy broadcasting is bidirectional but only for element-wise operations, Numpy mode for Broadcast operator is one-directional. Probably more clear will be using separate `AutoBroadcastType` enum for element-wise operations and for Broadcast op. @lazarevevgeny What do you think about it?

Reply:
Yes, this makes sence.

Reply:
But we should keep backward compatibility for the Broadcast op.

Reply:
And emphasise once again the difference that is also not covered with NUMPY mode currently is that the target shape rank for BIDIRECTIONAL model can be less than rank of input tensor.

Reply:
Probably these all discrepancies comes from the fact, as Mateusz already mentioned, originally nGraph autobroadcasting rules where developed for element-wise operations. @lazarevevgeny You're right, current nGraph NUMPY `AutoBroadcastType` works like bidirectional, with limitations you pointed. I guess that separating modes for autobroadcasting (element-wise ops) from `Broadcast` modes may be good solution.

Reply:
@diyessi any objections to separate `AutoBroadcastType` for `Broadcast` and element-wise operations?

Reply:
if current mode NUMPY works as bidirectional, can we treat that as bidirectional and add a separate UNIDIRECTIONAL mode? 
Also, by unidirectional, do you mean the numpy method `broadcast_to()` ?

Reply:
@gaurides
To sum up. As I understand:
The main difference between BIDIRECTIONAL and Numpy for broadcast op is that for BIDIRECTIONAL target_shape.size() can be less than input.size()
Please look to `Expand` spec which is exactly broadcast in BIDIRECTIONAL mode (https://github.com/onnx/onnx/blob/master/docs/Operators.md#Expand).
1. Numpy is de facto BIDIRECTIONAL but only for element-wise operations
2. Numpy used in Broadcast operator is not BIDIRECTIONAL according to specification
3. Numpy `broadcast_to` is not BIDIRECTIONAL, `numpy.array(input) * numpy.ones(shape)` is BIDIRECTIONAL

Reply:
I think that introduction `UNIDIRECTIONAL` ensure lack of `Broadcast:v3`-`Broadcast:v1` backward compatibility of need of changing `Broadcast:v1` spec

Reply:
@arogowie-intel @lazarevevgeny I've introduced separate type class for `Broadcast;v3`, please to re-review

Reply:
Done